### PR TITLE
Style changes in voxelization and preprocessing utilities

### PR DESCRIPTION
**Pull request by nachovizzo**
Continuation of #362 

Style changes on voxelization utilities:

- Move the voxel downsample function so it belongs to its family (VoxelUtils)
- minor code style changes in voxelization
- Remove `VoxelHash` struct favoring partial template specialization of `std::hash` so the end user doesn't care what's going on behind the scenes

### CODE REVIEW

File: cpp/kiss_icp/core/Preprocessing.hpp [hunk 1]
```diff
@@ -28,6 +28,9 @@
 #include <vector>
 
 namespace kiss_icp {
+/// Voxelize point cloud keeping the original coordinates
```
Comment (nachovizzo) [inline cpp/kiss_icp/core/Preprocessing.hpp:31]: One question to discuss now in this particular PR:

Assuming we get rid of `CorrectKITTIScan`, then this "Preprocessing" module will only have two functions:

```cpp
std::vector<Eigen::Vector3d> VoxelDownsample(const std::vector<Eigen::Vector3d> &frame,
                                             const double voxel_size);

std::vector<Eigen::Vector3d> Preprocess(const std::vector<Eigen::Vector3d> &frame,
                                        double max_range,
                                        double min_range);
```

The `Preprocess` function, which is only cropping (so we could change the name of it), is only consumed in the kiss-icp pipeline, so potentially, we could leave it there.

This makes me think `VoxelDownsample` is the only missing function here, and we can put it in our new and shiny `VoxelUtiils.hpp` module.

File: cpp/kiss_icp/core/VoxelUtils.hpp [hunk 2]
```diff
@@ -25,21 +25,25 @@
...
 }
+/// Voxelize a point cloud keeping the original coordinates
+std::vector<Eigen::Vector3d> VoxelDownsample(const std::vector<Eigen::Vector3d> &point_cloud,
+                                             const double voxel_size);
 
-struct VoxelHash {
-    size_t operator()(const Voxel &voxel) const {
+}  // namespace kiss_icp
+
+template <>
```
Comment (nachovizzo) [inline cpp/kiss_icp/core/VoxelUtils.hpp:43]: so we need to vote for partial template specialisation or custom type, it's 1-1 now, so it's @benemer choice now :)

Follow up of https://github.com/PRBonn/kiss-icp/pull/362#discussion_r1673016356

File: cpp/kiss_icp/core/Preprocessing.hpp (hunk 1 above)
Comment (nachovizzo) [reply cpp/kiss_icp/core/Preprocessing.hpp]: Moving this to a separate PR now https://github.com/PRBonn/kiss-icp/pull/369

File: cpp/kiss_icp/core/Preprocessing.hpp [hunk 3]
```diff
@@ -39,7 +38,4 @@ std::vector<Eigen::Vector3d> Preprocess(const std::vector<Eigen::Vector3d> &fram
 /// Originally introduced the calibration factor)
 std::vector<Eigen::Vector3d> CorrectKITTIScan(const std::vector<Eigen::Vector3d> &frame);
 
-/// Voxelize point cloud keeping the original coordinates
```
Comment (nachovizzo) [inline cpp/kiss_icp/core/Preprocessing.hpp:42]: we should also remove it then from the .cpp

File: cpp/kiss_icp/core/VoxelUtils.hpp (hunk 2 above)
Comment (benemer) [reply cpp/kiss_icp/core/VoxelUtils.hpp:43]: I agree with @nachovizzo that using the `VoxelHash` struct makes reading a bit easier. However, I am not sure how "accessible" this hash function is in the first place. For example, it is not clear where it is coming from and what those numbers are. Given this, I think I am okay with "hiding" it and using the partial specialization.

Comment (nachovizzo) [reply cpp/kiss_icp/core/VoxelUtils.hpp:43]: oki doki!

**Comment by nachovizzo** [comment]
Now ready to review and :ship:

**Review by benemer — APPROVED** [review]
Can be merged from my side :+1:

**Review by tizianoGuadagnino — APPROVED** [review]

**Pull request merged**